---
title: Integrating stimulated vs. control PBMC datasets to learn cell-type-specific responses
author: "hongc2@ccf.org"
date: "1/16/2020"
output: html_document
---

```{r setup, include=FALSE}
library(knitr)
knitr::opts_chunk$set(echo = TRUE, eval = TRUE, include = TRUE)
```
Change to the working directory!

Reference:
- https://satijalab.org/seurat/v3.1/immune_alignment.html
- https://satijalab.org/seurat/v3.1/integration.html#sctransform

This tutorial walks through an alignment of two groups of PBMCs from [Kang et al, 2017](https://www.nature.com/articles/nbt.4042). In this experiment, PBMCs were split into a stimulated and control group and the stimulated group was treated with interferon beta. The response to interferon caused cell type-specific gene expression changes that make a joint analysis of all the data difficult, with cell clustering both by stimulation condition and by cell type. Here, we demonstrate our integration strategy, as described in [Stuart and Butler et al, 2018](https://www.biorxiv.org/content/early/2018/11/02/460147), for performing integrated analyses to promote the identification of common cell types and enable comparative analyses. While this example demonstrates the integration of two datasets (conditions), these methods have been extended to multiple datasets. Refer to the [workflow](https://satijalab.org/seurat/pancreas_integration_label_transfer.html) that provides an example of integrating four pancreatic islet datasets.

### Integration goals
The following tutorial is designed to give you an overview of the kinds of comparative analyses on complex cell types that are possible using the Seurat integration procedure. Here, we address three main goals:

- Identify cell types that are present in both datasets
- Obtain cell type markers that are conserved in both control and stimulated cells
- Compare the datasets to find cell-type-specific responses to stimulation

### Workflow in summary
We will harmonize the Pearson residuals that are output from `SCTransform`. As demonstrated below, the workflow consists of the following steps:

- Create a list of Seurat objects to integrate
- More than one sample is compared. Thus considering library size factor, an explicit normalization is required
- Cell cycle evaluation is also considered for better comparison
- Perform `SCTransform` separately for each dataset
- Run the `PrepSCTIntegration` function on the object list
- Integrate datasets, and proceed with joint analysis

### Download Kang et. al. 2017 Seurat data (Raw read count)
- [Kang et. al. 2017](https://www.nature.com/articles/nbt.4042)
- Load necessary R libraries

```{r kang_data, echo=TRUE, eval=TRUE}
library(data.table)
library(ggplot2)
library(Seurat)

options(future.globals.maxSize = 4000 * 1024^2)

pkg <- "ifnb.SeuratData"
if( !is.element(pkg, .packages(all.available = TRUE)) ) {
    install.packages("https://seurat.nygenome.org/src/contrib/ifnb.SeuratData_3.0.0.tar.gz", repos = NULL, type = "source")
}
library(pkg,character.only = TRUE)

#load Kang data
data("ifnb")
```

### Preprocessing and Normalization
- Mitochondrial genes do not exist in the read count matrix and thus QC by mitochondria contamination is skipped.
- It is recommended to check the cell cycle phase before performing the sctransform method. Since the counts need to be comparable between cells and each cell has a different number of total UMIs, we do a rough normalization by dividing by total counts per cell and taking the natural log. This method isn't as accurate as the sctransform method that we will use ultimately to identify cell clusters, but it is sufficient to explore sources of variation in our data.

```{r ce_sctf, warning=FALSE}
load('data/cycle.rda')

#split into the original samples
ifnb.list <- SplitObject(ifnb, split.by = "stim")
ifnb.list <- lapply(X = ifnb.list, function(seu) {
    message("This run will take 5+ min ...")
    seu <- NormalizeData(seu, verbose = TRUE) #the normalization result will be stored into .data slot.
    seu <- CellCycleScoring(seu, g2m.features=g2m_genes, s.features=s_genes)
    seu <- SCTransform(seu,verbose = FALSE)
    return(seu)
})
```

### Feature Selection
Next, select features for downstream integration, and run `PrepSCTIntegration`, which ensures that all necessary Pearson residuals have been calculated.

```{r ce_fsel, warning=FALSE}
sc.features <- SelectIntegrationFeatures(object.list = ifnb.list)

ifnb.list <- PrepSCTIntegration(object.list = ifnb.list,
                                anchor.features = sc.features,
                                verbose=FALSE)
```

### Perform integration (Canonical Correlation Analysis)
Integration is a powerful method that uses these shared sources of greatest variation to identify shared subpopulations across conditions or datasets [Stuart and Bulter et al. (2018)]. The goal of integration is to ensure that the cell types of one condition/dataset align with the same cell types of the other conditions/datasets (e.g. control macrophages align with stimulated macrophages).

```{r cca, fig.align="center", out.width = "60%", eval=TRUE, fig.cap = 'Fig1. Stuart T and Butler A, et. al. Comprehensive integration of single cell data, bioRxiv 2018'}
include_graphics("data/integration.png")
```

Specifically, this integration method expects "correspondences" or shared biological states among at least a subset of single cells across the groups. The steps in the integration analysis are outlined in the figure below:

The different steps applied are as follows:

Perform canonical correlation analysis (CCA):

* CCA identifies shared sources of variation between the conditions/groups. It is a form of PCA, in that it identifies the greatest sources of variation in the data, but only if it is shared or conserved across the conditions/groups (using the 3000 most variant genes from each sample).

* This step roughly aligns the cells using the greatest shared sources of variation.

> NOTE: The shared highly variable genes are used because they are the most likely to represent those genes distinguishing the different cell types present.

* Identify anchors or mutual nearest neighbors (MNNs) across datasets (sometimes incorrect anchors are identified): MNNs can be thought of as 'best buddies'. For each cell in one condition:
    + The cell's closest neighbor in the other condition is identified based on gene expression values - it's the best buddy.
    + The reciprocal analysis is performed, and if the two cells are buddies in both directions, then those cells will be marked as anchors to 'anchor' the two datasets together.
    + "The difference in expression values between cells in an MNN pair provides an estimate of the batch effect, which is made more precise by averaging across many such pairs. A correction vector is obtained and applied to the expression values to perform batch correction." [Stuart and Bulter et al. (2018)].

* Filter anchors to remove incorrect anchors: Assess the similarity between anchor pairs by the overlap in their local neighborhoods (incorrect anchors will have low scores) - do the adjacent cells have best buddies that are adjacent to each other?

* Integrate the conditions/datasets:
    + Use anchors and corresponding scores to transform the cell expression values, allowing for the integration of the datasets (different samples, datasets, modalities)
    + NOTE: Transformation of each cell uses a weighted average of the two cells of each anchor across anchors of the datasets. Weights determined by cell similarity score (distance between cell and k nearest anchors) and anchor scores, so cells in the same neighborhood should have similar correction values.
    + If cell types are present in one dataset, but not the other, then the cells will still appear as a separate sample-specific cluster. Now, using our SCTransform object as input, let's perform the integration across conditions.

```{r ce_cca, warning=FALSE}
immune.anchors <- FindIntegrationAnchors(object.list = ifnb.list,
                                         normalization.method = "SCT",
                                         anchor.features = sc.features,
                                         verbose=FALSE)

immune.combined <- IntegrateData(anchorset = immune.anchors,
                                 normalization.method = "SCT",
                                 verbose=FALSE)
```

### Perform an integrated analysis (Visualization)
Now proceed with downstream analysis (i.e. visualization, clustering) on the integrated dataset. You can see that after integration, cells group by two conditions (control vs. a stimulated group). Cluster annotations to display are from the data we downloaded.

```{r umap.ctype}
#Let us delete ifnb.list to free up the memory space
rm(ifnb)
rm(ifnb.list)
rm(immune.anchors)

#Make sure that your default assay is 'integrated'
DefaultAssay(immune.combined) <- "integrated"

immune.combined <- RunPCA(immune.combined, verbose = FALSE)
immune.combined <- RunUMAP(immune.combined, dims = 1:20)

# immune.combined <- FindNeighbors(immune.combined, reduction = "pca", dims = 1:20)
# immune.combined <- FindClusters(immune.combined, resolution = 0.5)

plots <- DimPlot(immune.combined, group.by = c("stim","seurat_annotations"), combine = FALSE)

plots <- lapply(X = plots, FUN = function(x) {
  p <- x + theme(legend.position = "top")
  p <- p + guides(color = guide_legend(nrow = 4, byrow = TRUE, override.aes = list(size = 2.5)))
  })

CombinePlots(plots)
```

To visualize the two conditions side-by-side, we can use the split.by argument to show each condition colored by sample.

```{r, r.umap.seurat_clusters, warning=FALSE}

DimPlot(immune.combined, reduction = "umap", split.by = "stim", group.by = "seurat_annotations", label = TRUE) + NoLegend()

```

Let us save the R variable so that we can continue to work.
```{r save.ce_integ.rd}
wkd <- "out"
if (!file.exists(wkd)){dir.create(wkd)}
save(immune.combined, file = file.path(wkd,'01_immune_combined.rd'), compress = TRUE)
```

### Things to know
- Important R function and packages: `lapply` and `ggplot2`
- CCA
